### PR TITLE
RD-3119 Migrate backend DB models and migration scripts to TS

### PR DESCRIPTION
**Pull request by qooban**
## Description

Changes introduced within this PR:
* All DB models were migrated to TypeScript (I followed [TypeScript guidelines for Sequelize v6](https://sequelize.org/v6/manual/typescript.html#usage-of--code-sequelize-define--code-))
* All migration scripts were migrated to TypeScript
* Uses of DB models in backend routes/handlers were updated to take advantage of new types

Similar PR in Composer: https://github.com/cloudify-cosmo/cloudify-blueprint-composer/pull/982

The size of the PR is misleading (most of the changes comes from `package-lock.json`, which will be changed)

## Screenshots / Videos
N/A

## Checklist
- [x] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires a documentation update.
- [x] I added proper labels to this PR.

## Tests
[Stage-UI-System-Test #2180](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/2180/) (04.03.2022 8:00 CET) - PASSED
[Stage-UI-System-Test #2190](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/2190/tests) (07.03.2022 7:40 CET) - PASSED (except 2 test failures present also on `master` branch)

## Documentation
N/A

### CODE REVIEW

File: backend/db/models/ResourcesModel.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    type: ResourceType;
+    creator: string;
+    data: any;
+}
+type ResourcesCreationAttributes = Optional<ResourcesAttributes, 'creator' | 'data'>;
+interface ResourcesInstance extends Model<ResourcesAttributes, ResourcesCreationAttributes>, ResourcesAttributes {
+    readonly id: number;
+    readonly createdAt: Date;
+    readonly updatedAt: Date;
+}
```
Comment (qooban) [inline backend/db/models/ResourcesModel.ts:15]: This is an example how TypeScript can help in type-checking for sequelize models. It works properly with `migration` scripts when specific models are accessed directly.

This kind of types should be applied for all models, but that is probably something for another ticket in which it should also be verified that type-checking works for models indirect uses (in `backend/handlers` or `backend/routes` where they are accessed through `db` object). WDYT?

File: backend/db/models/UserAppsModel.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { DataTypes, Model, ModelDefined, Optional, Sequelize } from 'sequelize';
+import { MODE_COMMUNITY, MODE_CUSTOMER, MODE_MAIN, Mode } from '../../serverSettings';
+
+interface UserAppsAttributes {
+    username: string;
+    appDataVersion: number;
+    mode: Mode;
+    tenant: string;
+    appData: { pages: any[] };
```
Comment (qooban) [inline backend/db/models/UserAppsModel.ts:9]: I used `any[]` here as we don't have a way to reuse types between frontend and backend, yet. 
Probably something to think about and implement in separate ticket.

File: backend/migrations/20170425132017-4_0-init.ts [hunk 3]
```diff
@@ -26,12 +28,12 @@ function createClientConfigs(queryInterface, Sequelize) {
         })
         .then(() =>
             queryInterface.addIndex('ClientConfigs', ['managerIp'], {
-                indicesType: 'UNIQUE'
+                type: 'UNIQUE'
```
Comment (qooban) [inline backend/migrations/20170425132017-4_0-init.ts:31]: Unexpected breaking change happened between Sequelize v4 and v5. See: https://github.com/sequelize/sequelize/issues/10831. So this part didn't work for quite a long time, I think.

I'm adding a single comment, but there were many other uses of `indicesType` in other migration files as well.

File: backend/migrations/20171011082922-4_2-UserAppsRoleColumnRemoval.ts [hunk 4]
```diff
@@ -1,27 +1,29 @@
...
     },
 
-    down(queryInterface, Sequelize) {
+    down(queryInterface: QueryInterface, Sequelize: typeof sequelize) {
         return queryInterface
             .addColumn('UserApps', ROLE_COLUMN_NAME, {
                 type: Sequelize.ENUM,
                 values: ['admin', 'user'],
-                notNull: false,
+                allowNull: false,
```
Comment (qooban) [inline backend/migrations/20171011082922-4_2-UserAppsRoleColumnRemoval.ts:26]: There is no `notNull` key in options for `addColumn` and it seems to me that the intention was to not allow null (as there is `defaultValue` defined).

File: backend/migrations/20190401111659-5_0-MovePageResourcesToFiles.ts [hunk 5]
```diff
@@ -1,30 +1,33 @@
...
 
-module.exports = {
-    up: (queryInterface, Sequelize, logger) => {
+export const { up, down } = {
+    up: (queryInterface: QueryInterface, Sequelize: typeof sequelize, logger: Logger) => {
         return ResourcesModel(queryInterface.sequelize, Sequelize)
             .findAll({
                 where: { type: ResourceTypes.PAGE },
-                attributes: [['resourceId', 'id'], 'createdAt', 'updatedAt', 'creator'],
+                attributes: ['resourceId', 'createdAt', 'updatedAt', 'creator'],
```
Comment (qooban) [inline backend/migrations/20190401111659-5_0-MovePageResourcesToFiles.ts:21]: Without that change, type-checking didn't work on results.

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts [hunk 6]
```diff
@@ -50,18 +59,20 @@ module.exports = {
...
-            if (!_.isEmpty(pageData.layout)) {
+    down: (queryInterface: QueryInterface, Sequelize: typeof sequelize) => {
+        migrate(queryInterface, Sequelize, (pageData: PageData) => {
+            if (pageData.layout && pageData.layout.length > 0) {
                 let layoutSection = pageData.layout[0];
                 pageData[layoutSection.type] = layoutSection.content;
 
                 if (layoutSection.type === LAYOUT.WIDGETS) {
-                    layoutSection = _.nth(pageData.layout, 1) || {};
+                    layoutSection = _.nth(pageData.layout, 1) || { type: LAYOUT.WIDGETS, content: [] };
```
Comment (qooban) [inline backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts:69]: I'm not sure about this part of the code.

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts [hunk 7]
```diff
@@ -33,15 +42,15 @@ function migrate(queryInterface, Sequelize, pageProcessor) {
...
-module.exports = {
-    up: (queryInterface, Sequelize) =>
-        migrate(queryInterface, Sequelize, pageData => {
-            function migrateLayoutSection(layoutSection) {
-                if (pageData[layoutSection]) {
-                    if (!_.isEmpty(pageData[layoutSection])) {
-                        pageData.layout.push({ type: layoutSection, content: pageData[layoutSection] });
+export const { up, down } = {
+    up: (queryInterface: QueryInterface, Sequelize: typeof sequelize) =>
+        migrate(queryInterface, Sequelize, (pageData: PageData) => {
```
Comment (qooban) [inline backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts:47]: It's hard to add proper type-checking here as `pageData` is converted inside the function from `OldPageData` into `NewPageData`

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts [hunk 8]
```diff
@@ -50,18 +59,20 @@ module.exports = {
             migrateLayoutSection(LAYOUT.WIDGETS);
             migrateLayoutSection(LAYOUT.TABS);
         }),
-    down: (queryInterface, Sequelize) =>
-        migrate(queryInterface, Sequelize, pageData => {
-            if (!_.isEmpty(pageData.layout)) {
+    down: (queryInterface: QueryInterface, Sequelize: typeof sequelize) => {
+        migrate(queryInterface, Sequelize, (pageData: PageData) => {
```
Comment (qooban) [inline backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts:63]: It's hard to add proper type-checking here as `pageData` is converted inside the function from `NewPageData` into `OldPageData`

File: backend/migrations/20171011082922-4_2-UserAppsRoleColumnRemoval.ts (hunk 4 above)
Comment (kubama) [reply backend/migrations/20171011082922-4_2-UserAppsRoleColumnRemoval.ts:26]: It's 4.2 so probably it doesn't matter anyway, but if `notNull` did nothing then this changes the semantics. If so, some customers will have the constraint and some won't depending if they would run old or updated version of this migration. I don't think this is the way to go, such inconsistencies could lead to issues as I don't think anybody in the future will remember about it.

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts [hunk 9]
```diff
@@ -50,18 +59,20 @@ module.exports = {
...
+            if (pageData.layout && pageData.layout.length > 0) {
```
Comment (kubama) [inline backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts:64]: Can we use `?.` here?

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts (hunk 6 above)
Comment (kubama) [reply backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts:69]: In this case I would rather leave it the original implementation and make `type` and `content` props optional.

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts [hunk 10]
```diff
@@ -50,18 +59,20 @@ module.exports = {
...
 
                 if (layoutSection.type === LAYOUT.WIDGETS) {
-                    layoutSection = _.nth(pageData.layout, 1) || {};
+                    layoutSection = _.nth(pageData.layout, 1) || { type: LAYOUT.WIDGETS, content: [] };
                     if (layoutSection.type === LAYOUT.TABS) pageData[layoutSection.type] = layoutSection.content;
                 }
             }
 
+            // @ts-ignore Intentionally removing layout to convert pageData into OldPageData type
             delete pageData.layout;
```
Comment (kubama) [inline backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts:75]: Would `ts-ignore` be still required if `layout` prop was optional?

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts (hunk 8 above)
Comment (kubama) [reply backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts:63]: I think you handled it pretty well 👍

File: backend/migrations/20171011082922-4_2-UserAppsRoleColumnRemoval.ts (hunk 4 above)
Comment (qooban) [reply backend/migrations/20171011082922-4_2-UserAppsRoleColumnRemoval.ts]: Make sense, so I removed that line.

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts (hunk 9 above)
Comment (qooban) [reply backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts]: Yes, applied.

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts (hunk 6 above)
Comment (qooban) [reply backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts]: Done.

File: backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts (hunk 10 above)
Comment (qooban) [reply backend/migrations/20201030110539-5_1_1-MigratePageLayouts.ts]: No. Loosen the type and removed `ts-ignore`.

File: backend/consts.ts [hunk 11]
```diff
@@ -24,6 +24,6 @@ export const SERVER_HOST = 'localhost';
 export const SERVER_PORT = 8088;
 
 export const LAYOUT = {
-    TABS: 'tabs',
-    WIDGETS: 'widgets'
+    TABS: 'tabs' as const,
+    WIDGETS: 'widgets' as const
```
[the remark this answers is not in the document]
Comment (qooban) [reply backend/consts.ts:28]: To allow strict type-checking.

File: backend/db/models/UserAppsModel.ts (hunk 2 above)
Comment (qooban) [reply backend/db/models/UserAppsModel.ts:9]: Good point. I'll try to improve it.

Comment (qooban) [reply backend/db/models/UserAppsModel.ts]: Applied. Many type changes done, but seems like that was the right way to go 🙂

Comment (kubama) [reply backend/db/models/UserAppsModel.ts]: Looks cool, I appreciate the extra effort 👍.

**Review by kubama — APPROVED** [review]

**Review by kubama — APPROVED** [review]

**Pull request merged**